lib/hanami/settings.rb

```diff
@@ -125,9 +125,15 @@ def to_s
     def initialize(store = EMPTY_STORE)
       errors = config._settings.map(&:name).each_with_object({}) do |name, errs|
         value = store.fetch(name, Undefined)
-        next if value.eql?(Undefined)
 
-        public_send("#{name}=", value)
+        if value.eql?(Undefined)
+          # When a key is missing entirely from the store, _read_ its value from the config instead,
+          # which ensures its setting constructor runs (with a `nil` argument given) and raises any
+          # necessary errors.
+          public_send(name)
+        else
+          public_send("#{name}=", value)
+        end
```

Comment:
Actually, another way to do this might be just to assign the values only still, and then run `config.finalize!` at the end of this method, which would make it so the config cannot be subsequently written to, which is a positive side effect. I might try that in a few hours.